glide-core/redis-rs/redis/src/cluster_async/mod.rs

```diff
@@ -2337,6 +2453,58 @@ where
         }
     }
 
+    // The function is used to send an AUTH command to a node and then retry the original request.
+    // The function is used when the original request failed due to an AUTH error.
+    // Cases in which the function is helpful:
+    // 1. The password was changed, the connection was not re-established, and the request failed due to an AUTH error.
+    // 2. The protection method is allowing X hours of connection without authentication, and the request failed due to an AUTH error.
+    async fn send_re_auth_request(
+        core: Core<C>,
+        info: RequestInfo<C>,
+        address: String,
+    ) -> OperationResult {
+        let password = core
+            .get_cluster_param(|params| params.password.clone())
+            .expect(MUTEX_READ_ERR);
+        let username = core.get_cluster_param(|params| params.username.clone());
+        let Some(password) = password else {
+            return Err((
+                OperationTarget::Node { address },
+                RedisError::from((
+                    ErrorKind::AuthenticationFailed,
+                    "No password provided for AUTH",
+                )),
+            ));
+        };
+        let mut auth_cmd = crate::cmd("AUTH");
+        if let Ok(Some(username)) = username {
+            auth_cmd.arg(username);
+        }
+        auth_cmd.arg(password);
+        let cmd = Arc::new(auth_cmd.to_owned());
+        let routing =
+            InternalRoutingInfo::SingleNode(InternalSingleNodeRouting::ByAddress(address.clone()));
```

Comment:
hmmm, why authenticating only single connection and not all connections when we find an auth error?

Reply:
you're getting the address from outside - but the following request might be routed now to a different node if topology was changed in the meanwhile (the route would be set in try_request), so you'll reauth one node and then try to send the request to another.
Reautheticating all nodes should be the way to handle auth error when we find one - as the auth is per client - meaning all nodes should get the same password. I would actually use the recovery state to 'block' the entire pipeline of requests and set the password for all nodes (same as we do in recovery state for refresh slots or refresh connections), and then continue with the request. Have you considered it?

Reply:
There are two issues with what you offer -
1. NOAUTH error may happen on a simple disconnection from one node only, in such case blocking and sending to all nodes is super expensive.
2. The same as the topology changed in the fraction between reauth to resend, there's the chance of password changed but not all nodes are updated yet then ill fail all the requests I sent, and basically will lose the command.

Based on comparing both cases, I prefer to reauth in case of auth error, and in the rare case I was moved the reauth will happen again. If all nodes needing re-auth so whenever a request arrives to a node a reauth will be performed. I think the lazy method is smarter as it's not blocking the whole cluster for recovery but creates some small jumps here and there, at the end all request will complete without noticeable different behavior in latency or responses.

It's important to remember that we are comparing two rare cases, you offer topology change at a time of reauth I offer latency in cluster gossip.

I think the more significant is not to block and go to recovery, but to have a lazy handle.

Reply:
> 1. NOAUTH error may happen on a simple disconnection from one node only, in such case blocking and sending to all nodes is super expensive. 

When a connection gets disconnected it's being established again with the auth token stored in the client. If the stored auth token after the disconnection isn't the right one, we wouldn't get into this code - as the connection establishment would fail. If the stored password is correct - again, we wouldn't get into this block after a disconnection. So this case of disconnection isn't feasible. 
I think we should think through all possible scenarios to get this error and decide based on it.
2. You can decide to send auth to all nodes and not to fail it if some requests fail, then retry the request and if AUTH is still faulty it would return NO-AUTH error as it should.

More options if you want to keep the existing design by sending the AUTH only to a specific node:
1. pipeline both requests together (e.g with multi) so both would get directed to the same node. 
2. use the mechanism we're doing for ASK error. When ASK error is being received, we need to execute "ASKING" request on the connection before we run the request again. We do it in get_connection. It ensures that you run the AUTH on the relevant connection you're going to execute the command on. Again, if the AUTH request fails - I think you can log and ignore it and continue to retry to send the request, as if the auth wasn't updated on the client the original request would generate noauth error again which should be return to the user. However this option requires additional mechanism to make sure we don't do more than a single reauth try, so we won't get into a loop of auth errors.

But I think we should first describe all possible scenarios and then decide based on it